### PR TITLE
[RF] Pythonized RooFit tutorial files 

### PR DESCRIPTION
**Pull request by Harshalzzzzzzz**

### CODE REVIEW

File: bindings/pyroot/pythonizations/python/ROOT/pythonization/_roofit/_roosimultaneous.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+Some member functions of RooSimultaneous that take a RooCmdArg as argument also support keyword arguments.
+So far, this applies to RooSimultaneous::plotOn.
+For example, the following code is equivalent in PyROOT:
+\code{.py}
+# Directly passing a RooCmdArg:
+data.plotOn(frame, )
+
+# With keyword arguments:
+data.plotOn(frame,)
```
Comment (guitargeek) [inline bindings/pyroot/pythonizations/python/ROOT/pythonization/_roofit/_roosimultaneous.py:32]: There are some command arguments missing in this example. You can use the `Range` example like in `RooAbsPdf`, as this command arg can also be used for the RooSimultaneous (and you should call it `pdfSim` to highlight it's a RooSimultaneous pdf, not `data`):
```Python
# Directly passing a RooCmdArg:
pdfSim.fitTo(data, ROOT.RooFit.Range("r1"))
      
# With keyword arguments:
pdfSim.fitTo(data, Range="r1")
```

File: tutorials/roofit/rf107_plotstyles.py [hunk 2]
```diff
@@ -45,41 +45,31 @@
 # ---------------------------------------
 
 # Use sqrt(sum(weights^2)) error instead of Poisson errors
-data.plotOn(frame1, ROOT.RooFit.DataError(ROOT.RooAbsData.SumW2))
+data.plotOn(frame1, DataError = ROOT.RooAbsData.SumW2)
```
Comment (guitargeek) [inline tutorials/roofit/rf107_plotstyles.py:48]: Just a note for the future: there is an `enum` here, so we should implement a pythonization of `DataError`.

File: tutorials/roofit/rf107_plotstyles.py [hunk 3]
```diff
@@ -45,41 +45,31 @@
...
+gauss.plotOn(frame1, LineColor = ROOT.kRed)
 
 # Change line style to dashed
-gauss.plotOn(frame2, ROOT.RooFit.LineStyle(ROOT.kDashed))
+gauss.plotOn(frame2, LineStyle = ROOT.kDashed)
 
 # Filled shapes in green color
-gauss.plotOn(frame3, ROOT.RooFit.DrawOption("F"),
-             ROOT.RooFit.FillColor(ROOT.kOrange), ROOT.RooFit.MoveToBack())
+gauss.plotOn(frame3, ROOT.RooFit.MoveToBack(), DrawOption = "F", FillColor = ROOT.kOrange)
```
Comment (guitargeek) [inline tutorials/roofit/rf107_plotstyles.py:69]: Note for the future: we also need to do something about command arguments that don't take any parameters by default so we can do `MoveToBack=True` later.

File: tutorials/roofit/rf202_extendedmlfit.py [hunk 4]
```diff
@@ -68,22 +68,15 @@
 # rather than observed number of events (==data.numEntries())
 xframe = x.frame(ROOT.RooFit.Title("extended ML fit example"))
```
Comment (guitargeek) [inline tutorials/roofit/rf202_extendedmlfit.py:69]: The `frame` function is used so often, it's high on the priority lost for pythonizations for sure.

File: tutorials/roofit/rf205_compplot.py [hunk 5]
```diff
@@ -71,57 +71,34 @@
...
 # expressions on name
-model.plotOn(
-    xframe2,
-    ROOT.RooFit.Components("bkg1,sig*"),
-    ROOT.RooFit.LineStyle(
-        ROOT.kDashed),
-    ROOT.RooFit.LineColor(
-        ROOT.kYellow),
-    ROOT.RooFit.Invisible())
+model.plotOn(xframe2,  ROOT.RooFit.Invisible(), Components = "bkg1,sig*", LineStyle = ROOT.kDashed, LineColor= ROOT.kYellow)
```
Comment (guitargeek) [inline tutorials/roofit/rf205_compplot.py:101]: Actually, `Invisible` takes an argument, which is `true` by default:
https://github.com/root-project/root/blob/master/roofit/roofitcore/inc/RooGlobalFunc.h#L100

Just like in `Save`, by the way.

Please use `Invisible=True` here.

File: tutorials/roofit/rf209_anaconv.py [hunk 6]
```diff
@@ -45,7 +45,7 @@
 
 # Construct decay(t) (x) gauss1(t)
 decay_gm1 = ROOT.RooDecay("decay_gm1", "decay",
-                          dt, tau, gm1, ROOT.RooDecay.DoubleSided)
+                          dt, tau, gm1, DecayType = "DoubleSided")
```
Comment (guitargeek) [inline tutorials/roofit/rf209_anaconv.py:48]: This doesn't work yet, you will have to revert this change.

File: tutorials/roofit/rf606_nllerrorhandling.py [hunk 7]
```diff
@@ -82,10 +82,10 @@
 frame2 = m0.frame(ROOT.RooFit.Range(5.288, 5.293), ROOT.RooFit.Title("-log(L) scan vs m0, regions masked"))
 nll.plotOn(
     frame2,
-    ROOT.RooFit.PrintEvalErrors(-1),
     ROOT.RooFit.ShiftToZero(),
```
Comment (guitargeek) [inline tutorials/roofit/rf606_nllerrorhandling.py:85]: `ShiftToZero` is another one of these argument-less command args that we have to deal with, probably on the C++ side.

File: bindings/pyroot/pythonizations/python/ROOT/pythonization/_roofit/__init__.py [hunk 8]
```diff
@@ -22,10 +22,11 @@
 from ._rooarglist import RooArgList
 from ._rooargset import RooArgSet
 from ._rooworkspace import RooWorkspace
+from ._roosimultaneous import RooSimultaneous
 
 
 # list of python classes that are used to pythonize RooFit classes
-python_classes = [RooAbsCollection, RooAbsData, RooAbsPdf, RooAbsReal, RooArgList, RooArgSet, RooWorkspace]
+python_classes = [RooAbsCollection, RooAbsData, RooAbsPdf, RooAbsReal, RooArgList, RooArgSet, RooWorkspace, RooSimultaneous]
```
Comment (guitargeek) [inline bindings/pyroot/pythonizations/python/ROOT/pythonization/_roofit/__init__.py:29]: Please keep this list in alphabetic order (same for the includes above). This makes it easier for us to read the code later.

File: tutorials/roofit/rf402_datahandling.py [hunk 9]
```diff
@@ -144,8 +144,7 @@
 dh2.Print("v")
 
 # Add dh2 to yframe and redraw
-dh2.plotOn(yframe, ROOT.RooFit.LineColor(ROOT.kRed),
-           ROOT.RooFit.MarkerColor(ROOT.kRed))
+dh2.plotOn(yframe, LineColor = ROOT.kRed, MarkerColor = ROOT.kRed)
```
Comment (guitargeek) [inline tutorials/roofit/rf402_datahandling.py:147]: Here,  `dh2` is a RooDataHist, and `RooDataHist::plotOn` is not pythonized yet.

Either you add a pythonization for `RooDataHist::plotOn`, or you revert this change.

File: tutorials/roofit/rf708_bphysics.py [hunk 10]
```diff
@@ -64,47 +64,30 @@
...
 # Plot mixed slice for B0 and B0bar tagged data separately
 frame2 = dt.frame(ROOT.RooFit.Title(
     "B decay distribution of mixed events (B0/B0bar)"))
 
-data.plotOn(frame2, ROOT.RooFit.Cut(
-    "mixState==mixState::mixed&&tagFlav==tagFlav::B0"))
-bmix.plotOn(frame2, ROOT.RooFit.Slice(tagFlav, "B0"),
-            ROOT.RooFit.Slice(mixState, "mixed"))
+data.plotOn(frame2, Cut = "mixState==mixState::mixed&&tagFlav==tagFlav::B0")
+bmix.plotOn(frame2, Slice = (tagFlav, "B0"), Slice = (mixState, "mixed"))
```
Comment (guitargeek) [inline tutorials/roofit/rf708_bphysics.py:78]: You can't repeat keyword arguments, so we can't pythonize `Slice` yet. For now, you need to revert the changes to the `plotOn` calls where you pass the `Slice` command argument twice.

**Comment by guitargeek** [comment]
HI @Harshalzzzzzzz, thanks for making all these changes! I made some comments. Some comments request actual changes to make the tutorials run, some comments are highlighting some places where we need to implement new pythonizations in the future.

Some general comments:

1. Before making any PR, please make sure that none of the RooFit python tutorials are failing because of wrongly used pythonizations. I share with you [my script](https://github.com/guitargeek/gsoc-roofit-project-documentation/blob/main/2021_pythonization/scripts/run_roofit_tutorials.py) that you can use to automatically check if all tutorials are running fine (works both in python 2 and python 3).

2. In most python style conventions, there are no spaces around the `=` sign when you use keyword arguments. It would be good if you also followed this convention, so please do for example:
    ```Python 
    data.plotOn(frame, LineColor=ROOT.kRed) # instead of data.plotOn(frame, LineColor = ROOT.kRed)
    ```
    You don't have change all of this manually. What I recommend here is that we use an automatic code formatter for the tutorials. I suggest to use [black](https://github.com/psf/black) with the `--line-length=120` option.

    What you can do here is once you implemented the changes I suggested, you format the whole `tutorials/roofit` directory with:
    ```
    black --line-length=120 totorials/roofit
    ```
    Then you add a new commit with **only the formatting changes** that you give a commit message like "[RF] Format code of RooFit python tutorials". It's important that formatting changes and code changes are not mixed in the same commit. You will see that also many lines that you didn't touch before will get reformatted. And for all developments after that, we make sure to always run the code formatting before every new commit.

3. Your commits should be in a sensible order. Now, you have your pythonization of the tutorial files in the first commit, but this commit relies on the pythonization of RooSimultaneous that are only introduced in the second commit. But it's important that for every commit, all unit tests run. Since the tutorials are part of the continuous integration tests, they should run as well. This is important for finding possible bugs automatically with `git bisect`, for example.

    Can you please reorder the commits with `git rebase -i HEAD~2` such that the pythonization of RooSimultaneous happens before the changes to the tutorials?

In conclusion: once you reordered the commits, updated the changes to the tutorials such that everything runs, and added a new commit with the code formatting, you can go ahead and make the PR to the official ROOT repository.
